File: cli/context_builder.py
from budgets import BUDGETS
from models import Section, AssembleResult, RetrievedDoc
from tokenizer import estimate_tokens
from retrieval import retrieve


def truncate(text: str, max_tokens: int) -> tuple[str, bool]:
    tokens = estimate_tokens(text)
    if tokens <= max_tokens:
        return text, False

    ratio = max_tokens / tokens
    return text[: int(len(text) * ratio)], True


def assemble_context(
    query: str,
    memory: str,
    tool_outputs: str,
) -> AssembleResult:
    sections = {
        "instructions": Section(
            content="You are a helpful AI assistant. Follow system instructions strictly.",
            source="System",
            priority=1,
        ),
        "goal": Section(
            content=query,
            source="User Query",
            priority=2,
        ),
        "memory": Section(
            content=memory,
            source="Conversation Memory",
            priority=3,
        ),
        "retrieval": Section(
            content="",
            source="Vector DB",
            priority=4,
        ),
        "toolOutputs": Section(
            content=tool_outputs,
            source="Tool Outputs",
            priority=5,
        ),
    }

    # Truncate fixed sections
    for key in ["instructions", "goal", "memory", "toolOutputs"]:
        content, truncated = truncate(sections[key].content, BUDGETS[key])
        sections[key].content = content
        sections[key].truncated = truncated
        sections[key].budget_used = estimate_tokens(content)

    # Retrieval section
    retrieved_docs = retrieve(query)
    used_tokens = 0
    retrieval_text = ""
    meta: list[RetrievedDoc] = []

    for doc in retrieved_docs:
        block = f"[{doc.id.upper()}] {doc.title}\n{doc.content}\n"
        tokens = estimate_tokens(block)

        if used_tokens + tokens <= BUDGETS["retrieval"]:
            retrieval_text += block
            used_tokens += tokens
            doc.included = True
            doc.tokens_used = tokens
        else:
            doc.reason = "Budget exceeded"

        meta.append(doc)

    sections["retrieval"].content = retrieval_text
    sections["retrieval"].budget_used = used_tokens
    sections["retrieval"].details = meta
    sections["retrieval"].truncated = any(not d.included for d in meta)

    final_context = "\n---\n".join(
        sec.content for sec in sorted(sections.values(), key=lambda s: s.priority)
    )

    return AssembleResult(
        sections=sections,
        budgets=BUDGETS,
        final_context=final_context,
        retrieved_docs_meta=meta,
    )

File: ui/src/assets/lib/contextBuilder.ts
import { BUDGETS, enforceBudget, type BudgetKey } from './budget';
import {
  retrieveDocuments,
  type RetrievedDoc,
} from './retrieval';
import { estimateTokens } from './tokenizer';

/* ----------------------------- */
/* Section model                 */
/* ----------------------------- */
export interface Section {
  name: BudgetKey;
  content: string;
  source: string;
  priority: number;

  truncated: boolean;

  originalTokens: number;
  finalTokens: number;
  truncatedTokens: number;

  /* Retrieval-only metadata */
  details?: RetrievedDoc[];
}

/* ----------------------------- */
/* Assemble result               */
/* ----------------------------- */
export interface AssembleResult {
  sections: Section[];
  budgets: Record<BudgetKey, number>;
  finalContext: string;
}

/* ----------------------------- */
/* Context assembly              */
/* ----------------------------- */
export function assembleContext(
  query: string,
  memory = '',
  toolOutputs = ''
): AssembleResult {
  /* ----------------------------- */
  /* Base sections (pre-truncation) */
  /* ----------------------------- */
  const sections: Section[] = [
    {
      name: 'instructions',
      content:
        'You are an AI assistant specialized in machine learning and AI concepts. Answer accurately and concisely. If information is missing, say so explicitly.',
      source: 'System',
      priority: 1,
      truncated: false,
      originalTokens: 0,
      finalTokens: 0,
      truncatedTokens: 0,
    },
    {
      name: 'goal',
      content: query,
      source: 'User Query',
      priority: 2,
      truncated: false,
      originalTokens: 0,
      finalTokens: 0,
      truncatedTokens: 0,
    },
    {
      name: 'memory',
      content: memory,
      source: 'Conversation Memory',
      priority: 3,
      truncated: false,
      originalTokens: 0,
      finalTokens: 0,
      truncatedTokens: 0,
    },
    {
      name: 'retrieval',
      content: '',
      source: 'Vector Retrieval',
      priority: 4,
      truncated: false,
      originalTokens: 0,
      finalTokens: 0,
      truncatedTokens: 0,
    },
    {
      name: 'toolOutputs',
      content: toolOutputs,
      source: 'Recent Tool Outputs',
      priority: 5,
      truncated: false,
      originalTokens: 0,
      finalTokens: 0,
      truncatedTokens: 0,
    },
  ];

  /* ----------------------------- */
  /* Enforce budgets (non-retrieval) */
  /* ----------------------------- */
  for (const section of sections) {
    if (section.name === 'retrieval') continue;

    const originalTokens = estimateTokens(section.content);
    const enforced = enforceBudget(
      section.content,
      BUDGETS[section.name]
    );
    const finalTokens = estimateTokens(enforced.content);

    section.content = enforced.content;
    section.truncated = enforced.truncated;
    section.originalTokens = originalTokens;
    section.finalTokens = finalTokens;
    section.truncatedTokens = originalTokens - finalTokens;
  }

  /* ----------------------------- */
  /* Deterministic retrieval        */
  /* ----------------------------- */
  const retrievalResult = retrieveDocuments(query);
  const retrievalSection = sections.find(
    s => s.name === 'retrieval'
  )!;

  retrievalSection.content = retrievalResult.text;
  retrievalSection.details = retrievalResult.details;

  const totalRetrievalTokens = retrievalResult.details.reduce(
    (sum, doc) => sum + doc.tokens,
    0
  );

  retrievalSection.originalTokens = totalRetrievalTokens;
  retrievalSection.finalTokens = retrievalResult.tokensUsed;
  retrievalSection.truncatedTokens =
    totalRetrievalTokens - retrievalResult.tokensUsed;

  retrievalSection.truncated = retrievalResult.details.some(
    d => !d.included
  );

  /* ----------------------------- */
  /* Final context assembly         */
  /* ----------------------------- */
  const finalContext = sections
    .sort((a, b) => a.priority - b.priority)
    .map(s => s.content)
    .join('\n---\n');

  return {
    sections,
    budgets: BUDGETS,
    finalContext,
  };
}
